README.md
# Pools

[![Tests](https://github.com/AndreMiras/libpools/workflows/Tests/badge.svg?branch=develop)](https://github.com/AndreMiras/libpools/actions?query=workflow%3ATests)
[![Coverage Status](https://coveralls.io/repos/github/AndreMiras/libpools/badge.svg?branch=develop)](https://coveralls.io/github/AndreMiras/libpools?branch=develop)
[![PyPI version](https://badge.fury.io/py/pools.svg)](https://badge.fury.io/py/pools)

Python library for Ethereum pools liquidity providers.

## Install
```sh
pip install pools
```

## Usage
The `pools` library relies on `web3` which requires the `WEB3_INFURA_PROJECT_ID` environment variable to be set.
```sh
export WEB3_INFURA_PROJECT_ID=00000000000000000000000000000000
```
Then use the library to fetch portfolio data.
```python
>>> from pools import uniswap
>>> address = "0x000000000000000000000000000000000000dEaD"
>>> portfolio_data = uniswap.portfolio(address)
>>> portfolio_data.keys()
dict_keys(['address', 'pairs', 'balance_usd'])
>>> portfolio_data["balance_usd"]
Decimal('1234.56')
```
A Command line interface is also available.
```text
pools --help
Consider installing rusty-rlp to improve pyrlp performance with a rust based backend
usage: pools [-h] address

Liquidity provider portfolio stats.

positional arguments:
  address     Address

optional arguments:
  -h, --help  show this help message and exit
```


CHANGELOG.md
# Change Log


## [20210424]
  - Python 3.9 support
  - Setup coverage testing, refs #2
  - Auto publish to PyPI on release, refs #3


## [20201026]
  - Handle TheGraph `TransportQueryError`
  - Fix crash on total supply zero
  - Pulls token daily, refs #6


## [20201024]
  - Clean transaction types
  - Setup coverage testing, refs #2
  - Command line interface install, refs #5


## [20201023]
  - Setup CI, refs #1
  - Ship missing `abi.json`, refs #4


## [20201022]
  - Initial release


tests/utils.py
from unittest import mock

GQL_ETH_PRICE_RESPONSE = {"bundle": {"ethPrice": "321.123"}}
GQL_LIQUIDITY_POSITIONS_RESPONSE = {
    "user": {
        "liquidityPositions": [
            {
                "liquidityTokenBalance": "65.417152403305745713",
                "pair": {
                    "id": "0x3b3d4eefdc603b232907a7f3d0ed1eea5c62b5f7",
                    "reserve0": "98885.875625086259763385",
                    "reserve1": "3065.622053657196599417",
                    "reserveUSD": "2755342.621143665226669595853113687",
                    "token0": {
                        "derivedETH": "0.03100161710940527870014085576340626",
                        "id": "0x0ae055097c6d159879521c384f1d2123d1f195e6",
                        "name": "STAKE",
                        "symbol": "STAKE",
                    },
                    "token0Price": "32.25638186779036564112849328358329",
                    "token1": {
                        "derivedETH": "1",
                        "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "name": "Wrapped Ether",
                        "symbol": "WETH",
                    },
                    "token1Price": "0.03100161710940527870014085576340626",
                    "totalSupply": "12132.548610419336726782",
                },
            },
            {
                "liquidityTokenBalance": "123.321",
                "pair": {
                    "id": "0xd3d2e2692501a5c9ca623199d38826e513033a17",
                    "reserve0": "7795837.60970437134772868",
                    "reserve1": "64207.224033613483840543",
                    "reserveUSD": "48844843.23332099147592073020832003",
                    "token0": {
                        "derivedETH": "0.008236090494456606334236333082884844",
                        "id": "0x1f9840a85d5af5bf1d1762f925bdaddc4201f984",
                        "name": "Uniswap",
                        "symbol": "UNI",
                    },
                    "token0Price": "121.4168300692010713970072022761557",
                    "token1": {
                        "derivedETH": "1",
                        "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "name": "Wrapped Ether",
                        "symbol": "WETH",
                    },
                    "token1Price": "0.008236090494456606334236333082884844",
                    "totalSupply": "383443.946054848107867734",
                },
            },
        ]
    }
}

GQL_PAIR_INFO_RESPONSE = {
    "pair": {
        "id": "0xa478c2975ab1ea89e8196811f51a7b7ade33eb11",
        "reserve0": "202079477.297395245222385992",
        "reserve1": "554825.663433614212350256",
        "reserveUSD": "438900192.169828320338927756595308",
        "token0": {
            "derivedETH": "0.002745581445745187399781487618568183",
            "id": "0x6b175474e89094c44da98b954eedeac495271d0f",
            "name": "Dai Stablecoin",
            "symbol": "DAI",
        },
        "token0Price": "364.2215755608687365540815738979592",
        "token1": {
            "derivedETH": "1",
            "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "name": "Wrapped Ether",
            "symbol": "WETH",
        },
        "token1Price": "0.002745581445745187399781487618568183",
        "totalSupply": "8967094.518364383041536096",
    }
}


GQL_MINTS_BURNS_TX_RESPONSE = {
    "burns": [],
    "mints": [
        {
            "amount0": "15860000",
            "amount1": "600",
            "amountUSD": "229661.2283368789267441858327732994",
            "liquidity": "97549.987186057589967631",
            "pair": {"id": "0xf227e97616063a0ea4143744738f9def2aa06743"},
            "sender": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
            "to": "0x000000000000000000000000000000000000dead",
            "transaction": {
                "blockNumber": "11046485",
                "id": "0x7f9080f8c72c0ec21ec7e1690b94c52ebc4787bca66f2d154f6274...",
                "timestamp": "1602581467",
            },
        },
        {
            "amount0": "23188460.096098020166920577",
            "amount1": "1649.824913049740795957",
            "amountUSD": "531596.1714480471128118203674972062",
            "liquidity": "195593.709412655447555532",
            "pair": {"id": "0xc822d85d2dcedfaf2cefcf69dbd5588e7ffc9f10"},
            "sender": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
            "to": "0x000000000000000000000000000000000000dead",
            "transaction": {
                "blockNumber": "10543065",
                "id": "0x08d4f7eb1896d9ec25d2d36f72252cdb45f735b922fd1e515e1ce6...",
                "timestamp": "1595873620",
            },
        },
    ],
}


def patch_web3_contract(m_contract):
    return mock.patch("pools.uniswap.web3.eth.contract", m_contract)


def patch_client_execute(m_execute):
    return mock.patch("pools.uniswap.Client.execute", m_execute)


def patch_session_fetch_schema():
    """Bypassing `fetch_schema()` on unit tests."""
    return mock.patch("gql.client.SyncClientSession.fetch_schema")


pools/test_utils.py
from unittest import mock

GQL_ETH_PRICE_RESPONSE = {"bundle": {"ethPrice": "321.123"}}
GQL_LIQUIDITY_POSITIONS_RESPONSE = {
    "user": {
        "liquidityPositions": [
            {
                "liquidityTokenBalance": "65.417152403305745713",
                "pair": {
                    "id": "0x3b3d4eefdc603b232907a7f3d0ed1eea5c62b5f7",
                    "reserve0": "98885.875625086259763385",
                    "reserve1": "3065.622053657196599417",
                    "reserveUSD": "2755342.621143665226669595853113687",
                    "token0": {
                        "derivedETH": "0.03100161710940527870014085576340626",
                        "id": "0x0ae055097c6d159879521c384f1d2123d1f195e6",
                        "name": "STAKE",
                        "symbol": "STAKE",
                    },
                    "token0Price": "32.25638186779036564112849328358329",
                    "token1": {
                        "derivedETH": "1",
                        "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "name": "Wrapped Ether",
                        "symbol": "WETH",
                    },
                    "token1Price": "0.03100161710940527870014085576340626",
                    "totalSupply": "12132.548610419336726782",
                },
            },
            {
                "liquidityTokenBalance": "123.321",
                "pair": {
                    "id": "0xd3d2e2692501a5c9ca623199d38826e513033a17",
                    "reserve0": "7795837.60970437134772868",
                    "reserve1": "64207.224033613483840543",
                    "reserveUSD": "48844843.23332099147592073020832003",
                    "token0": {
                        "derivedETH": "0.008236090494456606334236333082884844",
                        "id": "0x1f9840a85d5af5bf1d1762f925bdaddc4201f984",
                        "name": "Uniswap",
                        "symbol": "UNI",
                    },
                    "token0Price": "121.4168300692010713970072022761557",
                    "token1": {
                        "derivedETH": "1",
                        "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "name": "Wrapped Ether",
                        "symbol": "WETH",
                    },
                    "token1Price": "0.008236090494456606334236333082884844",
                    "totalSupply": "383443.946054848107867734",
                },
            },
        ]
    }
}

GQL_PAIR_INFO_RESPONSE = {
    "pair": {
        "id": "0xa478c2975ab1ea89e8196811f51a7b7ade33eb11",
        "reserve0": "202079477.297395245222385992",
        "reserve1": "554825.663433614212350256",
        "reserveUSD": "438900192.169828320338927756595308",
        "token0": {
            "derivedETH": "0.002745581445745187399781487618568183",
            "id": "0x6b175474e89094c44da98b954eedeac495271d0f",
            "name": "Dai Stablecoin",
            "symbol": "DAI",
        },
        "token0Price": "364.2215755608687365540815738979592",
        "token1": {
            "derivedETH": "1",
            "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "name": "Wrapped Ether",
            "symbol": "WETH",
        },
        "token1Price": "0.002745581445745187399781487618568183",
        "totalSupply": "8967094.518364383041536096",
    }
}


GQL_MINTS_BURNS_TX_RESPONSE = {
    "burns": [],
    "mints": [
        {
            "amount0": "15860000",
            "amount1": "600",
            "amountUSD": "229661.2283368789267441858327732994",
            "liquidity": "97549.987186057589967631",
            "pair": {"id": "0xf227e97616063a0ea4143744738f9def2aa06743"},
            "sender": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
            "to": "0x000000000000000000000000000000000000dead",
            "transaction": {
                "blockNumber": "11046485",
                "id": "0x7f9080f8c72c0ec21ec7e1690b94c52ebc4787bca66f2d154f6274...",
                "timestamp": "1602581467",
            },
        },
        {
            "amount0": "23188460.096098020166920577",
            "amount1": "1649.824913049740795957",
            "amountUSD": "531596.1714480471128118203674972062",
            "liquidity": "195593.709412655447555532",
            "pair": {"id": "0xc822d85d2dcedfaf2cefcf69dbd5588e7ffc9f10"},
            "sender": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
            "to": "0x000000000000000000000000000000000000dead",
            "transaction": {
                "blockNumber": "10543065",
                "id": "0x08d4f7eb1896d9ec25d2d36f72252cdb45f735b922fd1e515e1ce6...",
                "timestamp": "1595873620",
            },
        },
    ],
}


GQL_TOKEN_DAY_DATA_RESPONSE = {
    "tokenDayDatas": [
        {"date": 1603584000, "priceUSD": "1.0037"},
        {"date": 1603497600, "priceUSD": "1.0053"},
        {"date": 1603411200, "priceUSD": "1.0063"},
        {"date": 1603324800, "priceUSD": "1.0047"},
        {"date": 1603238400, "priceUSD": "1.0059"},
        {"date": 1603152000, "priceUSD": "1.0049"},
    ]
}
GQL_PAIR_DAY_DATA_RESPONSE = {
    "pair": {
        "id": "0xa478c2975ab1ea89e8196811f51a7b7ade33eb11",
        "reserveUSD": "415905325.9588990528391949333277547",
        "token0": {
            "derivedETH": "0.002482164437276671900656302172320963",
            "id": "0x6b175474e89094c44da98b954eedeac495271d0f",
            "name": "Dai Stablecoin",
            "symbol": "DAI",
        },
        "token0Price": "402.87419519117702623465593526239",
        "token1": {
            "derivedETH": "1",
            "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "name": "Wrapped Ether",
            "symbol": "WETH",
        },
        "token1Price": "0.002482164437276671900656302172320963",
        "totalSupply": "8730969.742669688720211513",
    },
    "pairDayDatas": [
        {
            "date": 1603584000,
            "reserveUSD": "433176263.4363820888744425087438633",
            "totalSupply": "9069902.755513910975050686",
        },
        {
            "date": 1603497600,
            "reserveUSD": "435317156.2189432956087607791883648",
            "totalSupply": "9065803.30917003335268362",
        },
        {
            "date": 1603411200,
            "reserveUSD": "432572804.7594039159237288025989709",
            "totalSupply": "9033867.416922493126112964",
        },
        {
            "date": 1603324800,
            "reserveUSD": "431764296.8691012640837948223336187",
            "totalSupply": "8963586.802246124942226919",
        },
        {
            "date": 1603238400,
            "reserveUSD": "423735194.9935538342436108196570895",
            "totalSupply": "9037152.290229410449908895",
        },
        {
            "date": 1603152000,
            "reserveUSD": "411207002.0385470304033146757759173",
            "totalSupply": "9054265.837647610072038259",
        },
    ],
}

GQL_PAIRS_RESPONSE = {
    "pairs": [
        {
            "id": "0xc5ddc3e9d103b9dfdf32ae7096f1392cf88696f9",
            "reserve0": "2063243.37701238",
            "reserve1": "78990431.276124196481995237",
            "reserveUSD": "1155422539.501794978568848429540974",
            "token0": {
                "derivedETH": "1.384712347348822582084534991907731",
                "id": "0x4c6e796bbfe5eb37f9e3e0f66c009c8bf2a5f428",
                "name": "FC Bitcoin",
                "symbol": "FCBTC",
            },
            "token0Price": "0.02612016852775457641571125345392988",
            "token1": {
                "derivedETH": "0",
                "id": "0x975ce667d59318e13da8acd3d2f534be5a64087b",
                "name": "The Whale of Blockchain",
                "symbol": "TWOB",
            },
            "token1Price": "38.284592189266595295457534649458",
            "totalSupply": "6.764183030477266625",
        },
        {
            "id": "0xbb2b8038a1640196fbe3e38816f3e67cba72d940",
            "reserve0": "26186.56317714",
            "reserve1": "854243.645375842632389955",
            "reserveUSD": "688815654.2814067218630940203749505",
            "token0": {
                "derivedETH": "32.62144938979884788549711871554279",
                "id": "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
                "name": "Wrapped BTC",
                "symbol": "WBTC",
            },
            "token0Price": "0.03065467717423717613465000666387854",
            "token1": {
                "derivedETH": "1",
                "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "name": "Wrapped Ether",
                "symbol": "WETH",
            },
            "token1Price": "32.62144938979884788549711871554279",
            "totalSupply": "1.375359727911146499",
        },
        {
            "id": "0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc",
            "reserve0": "317611971.451732",
            "reserve1": "786437.873958944776984124",
            "reserveUSD": "634135172.5331979997924002078257594",
            "token0": {
                "derivedETH": "0.002476096446756450426416512921592668",
                "id": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
                "name": "USD//C",
                "symbol": "USDC",
            },
            "token0Price": "403.861489850261997342877776919223",
            "token1": {
                "derivedETH": "1",
                "id": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "name": "Wrapped Ether",
                "symbol": "WETH",
            },
            "token1Price": "0.002476096446756450426416512921592668",
            "totalSupply": "12.621500317891400641",
        },
    ]
}


def patch_web3_contract(m_contract):
    return mock.patch("pools.uniswap.web3.eth.contract", m_contract)


def patch_client_execute(m_execute):
    return mock.patch("pools.uniswap.Client.execute", m_execute)


def patch_session_fetch_schema():
    """Bypassing `fetch_schema()` on unit tests."""
    return mock.patch("gql.client.SyncClientSession.fetch_schema")


def patch_get_liquidity_positions(positions=None):
    positions = positions or []
    return mock.patch("pools.uniswap.get_liquidity_positions", return_value=positions)


def patch_get_staking_positions(positions=None):
    positions = positions or []
    return mock.patch("pools.uniswap.get_staking_positions", return_value=positions)


def patch_get_lp_transactions(mints_burns):
    return mock.patch("pools.uniswap.get_lp_transactions", return_value=mints_burns)


def patch_get_eth_price(price):
    return mock.patch("pools.uniswap.get_eth_price", return_value=price)


def patch_portfolio(data=None):
    data = data or {}
    return mock.patch("pools.uniswap.portfolio", return_value=data)


def patch_sys_argv(argv):
    return mock.patch("sys.argv", argv)


pools/uniswap.py
#!/usr/bin/env python
import argparse
import json
import os
from contextlib import contextmanager
from copy import deepcopy
from datetime import datetime
from decimal import Decimal, DivisionByZero, InvalidOperation
from pprint import pprint
from typing import Dict

from cachetools.func import ttl_cache
from gql import Client, gql
from gql.transport.exceptions import TransportQueryError, TransportServerError
from gql.transport.requests import RequestsHTTPTransport
from graphql import DocumentNode
from web3.auto.infura import w3 as web3

MODULE_DIRECTORY = os.path.dirname((os.path.abspath(__file__)))

# default cache settings
CACHE_MAXSIZE = 1000
CACHE_TTL = 5 * 60

# pool tokens that can be staked
# staking contract -> pool token
STAKING_POOLS = {
    # DAI-ETH
    "0xa1484C3aa22a66C62b77E0AE78E15258bd0cB711": "0xA478c2975Ab1Ea89e8196811F51A7B7Ade33eB11",  # noqa: E501
    # USDC-ETH
    "0x7FBa4B8Dc5E7616e59622806932DBea72537A56b": "0xB4e16d0168e52d35CaCD2c6185b44281Ec28C9Dc",  # noqa: E501
    # USDT-ETH
    "0x6C3e4cb2E96B01F4b866965A91ed4437839A121a": "0x0d4a11d5EEaaC28EC3F61d100daF4d40471f1852",  # noqa: E501
    # WBTC-ETH
    "0xCA35e32e7926b96A9988f61d510E038108d8068e": "0xBb2b8038a1640196FbE3e38816F3e67Cba72D940",  # noqa: E501
}

GQL_PAIR_PARAMETERS = """
id
token0 {
  id
  symbol
  name
  derivedETH
}
token1 {
  id
  symbol
  name
  derivedETH
}
reserve0
reserve1
totalSupply
reserveUSD
token0Price
token1Price
"""


class UniswapRoiException(Exception):
    """Base library exception."""


class InvalidAddressException(UniswapRoiException):
    """The Ethereum address is invalid."""


class TheGraphServiceDownException(UniswapRoiException):
    """When the graph is down, e.g. giving `502 Server Error: Bad Gateway`."""


@contextmanager
def gql_exceptions():
    try:
        yield
    except (TransportServerError, TransportQueryError) as e:
        raise TheGraphServiceDownException(e.args[0])


def get_gql_client():
    transport = RequestsHTTPTransport(
        url="https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2"
    )
    with gql_exceptions():
        return Client(transport=transport, fetch_schema_from_transport=True)


def gql_client_execute(document: DocumentNode, *args, **kwargs) -> Dict:
    client = get_gql_client()
    with gql_exceptions():
        return client.execute(document, *args, **kwargs)


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_eth_price():
    """Retrieves ETH price from TheGraph.com"""
    request_string = '{bundle(id: "1") {ethPrice}}'
    query = gql(request_string)
    result = gql_client_execute(query)
    eth_price = Decimal(result["bundle"]["ethPrice"])
    return eth_price


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_pair_info(contract_address):
    request_string = "query ($id: ID!) {pair(id: $id) {" + GQL_PAIR_PARAMETERS + "}}"
    query = gql(request_string)
    # note The Graph doesn't seem to like it in checksum format
    contract_address = contract_address.lower()
    variable_values = {"id": contract_address}
    result = gql_client_execute(query, variable_values=variable_values)
    return result


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_liquidity_positions(address):
    request_string = (
        """
        query ($id: ID!) {
          user(id: $id) {
            liquidityPositions (where: {liquidityTokenBalance_not: "0"}) {
              liquidityTokenBalance
              pair {
              """
        + GQL_PAIR_PARAMETERS
        + """
              }
            }
          }
        }
      """
    )
    query = gql(request_string)
    # note The Graph doesn't seem to like it in checksum format
    address = address.lower()
    variable_values = {"id": address}
    result = gql_client_execute(query, variable_values=variable_values)
    user = result["user"] or {}
    result = user.get("liquidityPositions", [])
    return result


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_staking_positions(address):
    """Given an address, returns all the staking positions."""
    # abi is the same for all contracts
    abi_path = os.path.join(MODULE_DIRECTORY, "abi.json")
    with open(abi_path) as f:
        abi = json.loads(f.read())
    positions = []
    for staking_contract, lp_contract in STAKING_POOLS.items():
        contract = web3.eth.contract(staking_contract, abi=abi)
        balance = contract.functions.balanceOf(address).call()
        if balance > 0:
            pair_info = get_pair_info(lp_contract)
            # this is the only missing key compared with the
            # `get_liquidity_positions()` call
            balance = web3.fromWei(balance, "ether")
            pair_info["liquidityTokenBalance"] = balance
            pair_info["pair"]["staking_contract_address"] = staking_contract
            positions.append(pair_info)
    return positions


def get_lp_transactions(address, pairs):
    """Retrieves mints/burns transactions of a given liquidity provider."""
    gql_order_by = "orderBy: timestamp, orderDirection: desc"
    gql_transaction_parameters = "transaction { id timestamp blockNumber }"
    gql_pair_parameters = "pair { id }"
    gql_mints_burns_parameters = "to sender liquidity amount0 amount1 amountUSD"
    gql_parameters = (
        gql_transaction_parameters
        + " "
        + gql_pair_parameters
        + " "
        + gql_mints_burns_parameters
    )
    request_string = (
        """
        query ($address: Bytes! $pairs: [String!]) {
          mints(
            where: { to: $address pair_in: $pairs}, """
        + gql_order_by
        + """
          ) {
        """
        + gql_parameters
        + """
          }
          burns(
            where: { sender: $address pair_in: $pairs}, """
        + gql_order_by
        + """
          ) {
        """
        + gql_parameters
        + """
          }
        }
      """
    )
    query = gql(request_string)
    # note The Graph doesn't seem to like it in checksum format
    address = address.lower()
    variable_values = {"address": address, "pairs": pairs}
    result = gql_client_execute(query, variable_values=variable_values)
    return result


def extract_pair_info(pair, balance, eth_price):
    """Builds a dictionary with pair information."""
    contract_address = None
    pair_symbol = None
    total_supply = None
    share = None
    tokens = []
    if pair:
        contract_address = pair["id"]
        # this was populated via `get_staking_positions()`
        staking_contract_address = pair.get("staking_contract_address")
        total_supply = Decimal(pair["totalSupply"])
        share = 100 * (balance / total_supply)
        reserve_usd = Decimal(pair["reserveUSD"])
        pool_token_price = reserve_usd / total_supply
        for i in range(2):
            token = pair[f"token{i}"]
            token_symbol = token["symbol"]
            token_price = Decimal(token["derivedETH"]) * eth_price
            token_balance = Decimal(pair[f"reserve{i}"]) * share * Decimal("0.01")
            token_balance_usd = token_balance * token_price
            tokens.append(
                {
                    "symbol": token_symbol,
                    "price_usd": token_price,
                    "balance": token_balance,
                    "balance_usd": token_balance_usd,
                }
            )
        pair_symbol = "-".join([token["symbol"] for token in tokens])
        balance_usd = sum([token["balance_usd"] for token in tokens])
    pair_info = {
        "contract_address": contract_address,
        "staking_contract_address": staking_contract_address,
        "owner_balance": balance,
        "symbol": pair_symbol,
        "total_supply": total_supply,
        "price_usd": pool_token_price,
        "share": share,
        "balance_usd": balance_usd,
        "tokens": tokens,
    }
    return pair_info


def merge_lp_transactions(mints, burns):
    """Merges mints/burns transactions and order by timestamp."""
    mints = list(map(lambda m: dict(m, **{"type": "mint"}), mints))
    burns = list(map(lambda m: dict(m, **{"type": "burn"}), burns))
    transactions = mints + burns
    # TODO: reverse=True
    transactions = sorted(transactions, key=lambda tx: tx["transaction"]["timestamp"])
    return transactions


def group_lp_transactions(transactions):
    """Groups transactions by pair."""
    transaction_dict = {}
    for transaction in transactions:
        pair = transaction["pair"]["id"]
        transaction_dict[pair] = transaction_dict.get(pair, []) + [transaction]
    return transaction_dict


def fix_type_lp_transactions(transactions):
    """Makes sure the type of each fields is correct."""
    transactions = deepcopy(transactions)
    for transaction in transactions:
        transaction["amount0"] = Decimal(transaction["amount0"])
        transaction["amount1"] = Decimal(transaction["amount1"])
        transaction["amountUSD"] = Decimal(transaction["amountUSD"])
        transaction["liquidity"] = Decimal(transaction["liquidity"])
        transaction_transaction = transaction["transaction"]
        block_number = transaction_transaction.pop("blockNumber")
        transaction_transaction["block_number"] = int(block_number)
        timestamp = transaction_transaction["timestamp"]
        transaction_transaction["timestamp"] = datetime.utcfromtimestamp(int(timestamp))
    return transactions


def clean_transactions(mints_burns):
    """
    Fixes ordering, grouping, typing and dict key convention of the mints burns result.
    """
    mints = mints_burns["mints"]
    burns = mints_burns["burns"]
    transactions = merge_lp_transactions(mints, burns)
    transactions = fix_type_lp_transactions(transactions)
    transaction_dict = group_lp_transactions(transactions)
    return transaction_dict


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def portfolio(address):
    try:
        address = web3.toChecksumAddress(address)
    except ValueError:
        raise InvalidAddressException(address)
    # TODO: check if the GraphQL queries can be merged into one
    eth_price = get_eth_price()
    positions = []
    positions += get_liquidity_positions(address)
    positions += get_staking_positions(address)
    pair_addresses = [pair["pair"]["id"] for pair in positions]
    mints_burns = get_lp_transactions(address, pair_addresses)
    transaction_dict = clean_transactions(mints_burns)
    balance_usd = 0
    pairs = []
    for position in positions:
        balance = Decimal(position["liquidityTokenBalance"])
        pair = position["pair"]
        pair_info = extract_pair_info(pair, balance, eth_price)
        contract_address = pair_info["contract_address"]
        # even though the pair exists, we may not find transactions history
        pair_info["transactions"] = transaction_dict.get(contract_address, [])
        balance_usd += pair_info["balance_usd"]
        pairs.append(pair_info)
    data = {
        "address": address,
        "pairs": pairs,
        "balance_usd": balance_usd,
    }
    return data


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_token_daily_raw(address):
    """
    Raw pull of token daily data from TheGraph.
    Note this getting the daily for the token, not the pair.
    """
    request_string = """
        query ($token: String!) {
          tokenDayDatas(
            orderBy: date,
            orderDirection: desc,
            first: 31,
            where: {token: $token}
          ) {
            date
            priceUSD
          }
        }
      """
    query = gql(request_string)
    # note The Graph doesn't seem to like it in checksum format
    address = address.lower()
    variable_values = {"token": address}
    result = gql_client_execute(query, variable_values=variable_values)
    result = result["tokenDayDatas"]
    return result


def fix_type_token_daily(data):
    """Makes sure the type of each fields is correct."""
    data = deepcopy(data)
    for data_day in data:
        price_usd = data_day.pop("priceUSD")
        data_day["price_usd"] = Decimal(price_usd)
        data_day["date"] = datetime.utcfromtimestamp(int(data_day["date"]))
    return data


def get_token_daily(address):
    data = get_token_daily_raw(address)
    data = fix_type_token_daily(data)
    return data


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_pair_daily_raw(address):
    """Raw pull of pair daily data from TheGraph."""
    request_string = (
        """
        query ($pairAddress: Bytes!, $id: ID!) {
          pair(id: $id) {"""
        + GQL_PAIR_PARAMETERS
        + """}
          pairDayDatas(
            orderBy: date,
            orderDirection: desc,
            first: 31,
            where: {pairAddress: $pairAddress}
          ) {
            date
            totalSupply
            reserveUSD
          }
        }
      """
    )
    query = gql(request_string)
    # note The Graph doesn't seem to like it in checksum format
    address = address.lower()
    variable_values = {"id": address, "pairAddress": address}
    result = gql_client_execute(query, variable_values=variable_values)
    pair = result["pair"]
    date_price = result["pairDayDatas"]
    result = {"pair": pair, "date_price": date_price}
    return result


def fix_type_pair_daily(data):
    """Makes sure the type of each fields is correct."""
    data = deepcopy(data)
    for data_day in data:
        reserve_usd = Decimal(data_day.pop("reserveUSD"))
        total_supply = Decimal(data_day.pop("totalSupply"))
        try:
            data_day["price_usd"] = reserve_usd / total_supply
        except (InvalidOperation, DivisionByZero):
            data_day["price_usd"] = Decimal(0)
        data_day["date"] = datetime.utcfromtimestamp(int(data_day["date"]))
    return data


def fix_pair(pair):
    pair = deepcopy(pair)
    total_supply = Decimal(pair.pop("totalSupply"))
    pair["total_supply"] = total_supply
    reserve_usd = Decimal(pair.pop("reserveUSD"))
    pair["reserve_usd"] = reserve_usd
    pair_price_usd = reserve_usd / total_supply
    pair["price_usd"] = pair_price_usd
    pair_symbol = pair["token0"]["symbol"] + "-" + pair["token1"]["symbol"]
    pair["symbol"] = pair_symbol
    return pair


def get_pair_daily(address):
    data = get_pair_daily_raw(address)
    # makes sure we don't mutate the cached reference
    # or next cache hit we would retrieve that already mutated reference
    data = deepcopy(data)
    pair = data["pair"]
    data["pair"] = fix_pair(pair)
    date_price = data["date_price"]
    data["date_price"] = fix_type_pair_daily(date_price)
    return data


@ttl_cache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL)
def get_pairs_raw():
    """Raw pull of pairs data from TheGraph."""
    request_string = (
        """
    {
     pairs(first: 10, orderBy: reserveUSD, orderDirection: desc) {"""
        + GQL_PAIR_PARAMETERS
        + """
     }
    }
    """
    )
    query = gql(request_string)
    result = gql_client_execute(query)
    result = result["pairs"]
    return result


def get_pairs():
    pairs = get_pairs_raw()
    pairs = deepcopy(pairs)
    for i, pair in enumerate(pairs):
        pairs[i] = fix_pair(pair)
    return pairs


def main():
    parser = argparse.ArgumentParser(description="Liquidity provider portfolio stats.")
    parser.add_argument("address", help="Address")
    args = parser.parse_args()
    address = args.address
    data = portfolio(address)
    pprint(data)
    return data


if __name__ == "__main__":
    main()  # pragma: no cover
